### PR TITLE
Content API v2 date formatting

### LINKED ISSUE #10065: Content API V2: Date formatting
refs #9866

### Why do we need yet another way of date format in our API?

We want to avoid unnecessary work on the client doing date conversions. At the moment steps would be:

1. Make a request to another endpoint to get the time zone
2. Load utilities like `moment-timezone` (which is big overhead compared to just loading `moment`) 
3. Know how to use that to format the date correctly

All these steps can be obliterated once we return formatted date directly from the API.

### How do we solve this problem?

Based on our research and internal conversations, the best solution for this problem would be adding [UTC offset](https://en.wikipedia.org/wiki/UTC_offset) information to an existing date field, that way it will follow current [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) format. For blog instance that is configured with `Europe/Oslo` timezone, example response would contain information like:
```

"updated_at": "2014-01-01T00:58:58+01:00"

```
comparing to currently returned:
```
"updated_at": "2013-12-31T23:58:58Z"
```
The implementation of this additional field should be solved in a similar fashion to the one here https://github.com/TryGhost/Ghost/issues/10062

### PR DESCRIPTION
**Pull request by naz**
closes #10065

- Added UTC offset to dates returned by Content API
- Added test checking new format is compatible with Admin API
- Refactored output serializer mapping logic

### CODE REVIEW

File: core/server/api/v2/utils/index.js [hunk 1]
```diff
@@ -9,5 +11,9 @@ module.exports = {
 
     get validators() {
         return require('./validators');
+    },
+
+    isContentAPI: (frame) => {
```
Comment (gargol) [inline core/server/api/v2/utils/index.js]: This refactoring happened before I dug out that this context check is not working any more passed `permissions` stage. Should probably be put in separate PR to be absolutely clean

File: core/server/api/v2/utils/serializers/output/utils/date.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+const moment = require('moment-timezone');
+const settingsCache = require('../../../../../../services/settings/cache');
+
+const format = (date) => {
+    return moment(date)
+        .tz(settingsCache.get('active_timezone'))
+        .toISOString(true);
```
Comment (gargol) [inline core/server/api/v2/utils/serializers/output/utils/date.js]: These 3 lines above are everything that this PR is meant to do. The output is an ISO date with offset information - https://momentjs.com/docs/#/displaying/as-iso-string/

File: core/server/api/v2/utils/serializers/output/pages.js [hunk 3]
```diff
@@ -1,21 +1,37 @@
 const debug = require('ghost-ignition').debug('api:v2:utils:serializers:output:pages');
 const url = require('./utils/url');
+const date = require('./utils/date');
+
+const mapPage = (model, frame) => {
+    const jsonModel = model.toJSON(frame.options);
+    url.forPost(model.id, jsonModel, frame.options);
+
+    if (frame.options.context.public) {
+        ['created_at', 'updated_at', 'published_at'].forEach((field) => {
```
Comment (gargol) [inline core/server/api/v2/utils/serializers/output/pages.js]: @kirrg001 just as discussed, simple iteration over known publicly exposed fields here, but definitely could be transformed to something more generic with iteration over all object fields and transforming it if it's a date for example.

**Comment by gargol** [comment]
@kirrg001 still working on more unit test for output serializers, but the main concept is present. 
cc @ErisDS in case you are interested :wink:

**Comment by ErisDS** [comment]
Off the top of my head: there’s one more date field on the user model something like last seen.

**Comment by gargol** [comment]
@ErisDS that field (`last_seen`), is never exposed to the public along with other datetime fields on user resource - https://github.com/TryGhost/Ghost/blob/master/core/server/models/user.js#L20e.

File: core/server/api/v2/utils/index.js (hunk 1 above)
Comment (kirrg001) [inline core/server/api/v2/utils/index.js]: -> (!user AND api_key) || public

Comment (kirrg001) [reply core/server/api/v2/utils/index.js]: https://github.com/TryGhost/Ghost/issues/10099

File: core/server/api/v2/utils/serializers/output/pages.js [hunk 4]
```diff
@@ -1,21 +1,38 @@
+const _ = require('lodash');
 const debug = require('ghost-ignition').debug('api:v2:utils:serializers:output:pages');
 const url = require('./utils/url');
+const date = require('./utils/date');
+
+const mapPage = (model, frame) => {
+    const jsonModel = model.toJSON(frame.options);
+    url.forPost(model.id, jsonModel, frame.options);
```
Comment (kirrg001) [inline core/server/api/v2/utils/serializers/output/pages.js]: Either we raise an issue or we change this behaviour as separate commit.

modification by ref vs modification by copy (we mix both)

File: core/server/api/v2/utils/serializers/output/pages.js [hunk 5]
```diff
@@ -1,21 +1,38 @@
...
+
+    if (_.get(frame, 'options.context.public')) {
+        ['created_at', 'updated_at', 'published_at'].forEach((field) => {
+            if (jsonModel[field]) {
+                jsonModel[field] = date.format(jsonModel[field]);
+            }
+        });
+    }
+
+    return jsonModel;
```
Comment (kirrg001) [inline core/server/api/v2/utils/serializers/output/pages.js]: No need to return if we modify by ref

File: core/test/functional/api/v2/admin/posts_spec.js [hunk 6]
```diff
@@ -415,6 +414,35 @@ describe('Posts API V2', function () {
...
+                    .set('Origin', config.get('url'))
+                    .send({posts: [post]})
+                    .expect('Content-Type', /json/)
+                    .expect('Cache-Control', testUtils.cacheRules.private)
+                    .expect(201)
+                    .then((res) => {
+                        res.body.posts.length.should.eql(1);
+                        testUtils.API.checkResponse(res.body.posts[0], 'post');
+                        should.not.exist(res.headers['x-cache-invalidate']);
+                        console.log(res.body.posts[0]);
```
Comment (kirrg001) [inline core/test/functional/api/v2/admin/posts_spec.js]: whops 🤠

**Comment by gargol** [comment]
@kirrg001 have pushed out a change to how relations are handled during mapping and fixed nested tag dates not being transformed at the same time. Still, have to redo some of the unit tests to conform to this new structure. Let me know if you have comments around this new way of organizing 'decoration' process :smiley:

File: core/server/api/v2/utils/serializers/output/utils/mapper.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+    if (utils.isContentAPI(frame)) {
+        date.forTag(jsonModel);
+    }
+
+    return jsonModel;
+};
+
+module.exports.mapPost = mapPost;
+module.exports.mapTag = mapTag;
```
Comment (kirrg001) [inline core/server/api/v2/utils/serializers/output/utils/mapper.js]: the users serializer should also move up it's logic to the mapper, otherwise it's confusing & inconsistent

File: core/server/api/v2/utils/serializers/output/utils/mapper.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    if (frame.options && frame.options.withRelated) {
+        frame.options.withRelated.forEach((relation) => {
+            // @NOTE: this block also decorates primary_tag/primary_author objects as they
+            // are being passed by reference in tags/authors. Might be refactored into more explicit call
+            // in the future, but is good enough for current use-case
+            if (relation === 'tags' && jsonModel.tags) {
+                jsonModel.tags = jsonModel.tags.map(tag => url.forTag(tag.id, tag));
+
+                if (utils.isContentAPI(frame)) {
+                    jsonModel.tags = jsonModel.tags.map(tag => date.forTag(tag));
```
Comment (kirrg001) [inline core/server/api/v2/utils/serializers/output/utils/mapper.js]: I am getting `tags=[null,null]` while testing.
This is probably because `forTag` does not return anything.

**Review by kirrg001 — CHANGES_REQUESTED** [review]
-

File: core/server/api/v2/utils/serializers/output/utils/mapper.js [hunk 9]
```diff
@@ -0,0 +1,52 @@
+const utils = require('../../../');
+const url = require('./url');
+const date = require('./date');
+
+const mapPost = (model, frame) => {
+    const jsonModel = model.toJSON(frame.options);
+
+    url.forPost(model.id, jsonModel, frame.options);
```
[the remark this answers is not in the document]
Comment (kirrg001) [reply core/server/api/v2/utils/serializers/output/utils/mapper.js]: IMO i would just make ^ this as an example. the decorator logic is in general hard to read i find. But just a personal impression :)

File: core/server/api/v2/utils/serializers/output/utils/mapper.js (hunk 7 above)
Comment (gargol) [reply core/server/api/v2/utils/serializers/output/utils/mapper.js]: :100: agree

File: core/server/api/v2/utils/serializers/output/utils/mapper.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (gargol) [reply core/server/api/v2/utils/serializers/output/utils/mapper.js]: Sure, raised an issue for later improvements - https://github.com/TryGhost/Ghost/issues/10102. Definitely need to come up with one way of modifying state, as it's getting messy pretty quickly :)

**Comment by kirrg001** [comment]
@gargol Please self merge. We will do QA together before releasing.

**Review by kirrg001 — APPROVED** [review]
Self merge 👍
